### PR TITLE
ENH: tilde expansion for write functions, #11438

### LINKED ISSUE #11438: Tilde expansion for paths (~/...) inconsistent between read_csv and to_csv
When specifying a path in DataFrame.read_csv(), pandas has no problem with the tilde expansion:

``` python
In[1]:
    import pandas as pd
    test = pd.read_csv('~/Documents/test.csv', index_col=0)
    test.head()

Out[1]:
        Foo Bar
    0   1   2
    1   3   4
    2   5   6
    3   7   8
    4   9   10
```

But DataFrame.to_csv() doesn't understand this:

``` python
In[2]:
    test.to_csv('~/Documents/test1.csv')

Out[2]:
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-4-ef1160d6e930> in <module>()
----> 1 test.to_csv("~/Documents/test1.csv")

//anaconda/lib/python3.4/site-packages/pandas/core/frame.py in to_csv(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, quoting, quotechar, line_terminator, chunksize, tupleize_cols, date_format, doublequote, escapechar, decimal, **kwds)
   1287                                      escapechar=escapechar,
   1288                                      decimal=decimal)
-> 1289         formatter.save()
   1290 
   1291         if path_or_buf is None:

//anaconda/lib/python3.4/site-packages/pandas/core/format.py in save(self)
   1467         else:
   1468             f = com._get_handle(self.path_or_buf, self.mode,
-> 1469                                 encoding=self.encoding)
   1470             close = True
   1471 

//anaconda/lib/python3.4/site-packages/pandas/core/common.py in _get_handle(path, mode, encoding, compression)
   2865                 f = open(path, mode, encoding=encoding)
   2866             else:
-> 2867                 f = open(path, mode, errors='replace')
   2868         else:
   2869             f = open(path, mode)

FileNotFoundError: [Errno 2] No such file or directory: '~/Documents/test1.csv'
```

Instead, it needs the file name alone, or the full path: `test.to_csv('/Users/alexchase/Documents/test1.csv')`.

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.4.3.final.0
python-bits: 64
OS: Darwin
OS-release: 15.0.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8

pandas: 0.17.0
nose: 1.3.7
pip: 7.1.2
setuptools: 18.4
Cython: 0.22.1
numpy: 1.10.1
scipy: 0.15.1
statsmodels: 0.6.1
IPython: 4.0.0
sphinx: 1.3.1
patsy: 0.3.0
dateutil: 2.4.2
pytz: 2015.6
blosc: None
bottleneck: 1.0.0
tables: 3.2.0
numexpr: 2.4.3
matplotlib: 1.4.3
openpyxl: 1.8.5
xlrd: 0.9.3
xlwt: 1.0.0
xlsxwriter: 0.7.3
lxml: 3.4.4
bs4: 4.3.2
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.5
pymysql: None
psycopg2: None
```

### PR DESCRIPTION
**Pull request by aechase**
closes #11438 
Added calls to `_expand_user()` in multiple Formatter classes so that writer functions, such as `to_csv()` and `to_html()`, can perform tilde expansion on file paths beginning `~/`. This makes the writer functions consistent with the reader functions, such as `read_csv()`. I also changed the `__init__()` for `CSVFormatter` to conform to the pep8 width limit.

### CODE REVIEW

**Comment by jreback** [comment]
tests pls

File: pandas/core/format.py [hunk 1]
```diff
@@ -68,12 +68,14 @@
...
 docstring_to_string = common_docstring + justify_docstring + return_docstring
 
+
 class CategoricalFormatter(object):
 
     def __init__(self, categorical, buf=None, length=True,
                  na_rep='NaN', footer=True):
         self.categorical = categorical
         self.buf = buf if buf is not None else StringIO(u(""))
+        self.buf = _expand_user(self.buf)
```
Comment (jreback) [inline pandas/core/format.py]: doesn't belong here

File: pandas/core/format.py [hunk 2]
```diff
@@ -130,6 +132,7 @@ def __init__(self, series, buf=None, length=True, header=True,
                  max_rows=None):
         self.series = series
         self.buf = buf if buf is not None else StringIO()
+        self.buf = _expand_user(self.buf)
```
Comment (jreback) [inline pandas/core/format.py]: eoesn't belong here

File: pandas/core/format.py [hunk 3]
```diff
@@ -923,6 +927,7 @@ def write_tr(self, line, indent=0, indent_delta=4, header=False,
     def write_result(self, buf):
         indent = 0
         frame = self.frame
+        buf = _expand_user(buf)
```
Comment (jreback) [inline pandas/core/format.py]: doesn't belong here

File: pandas/core/format.py [hunk 4]
```diff
@@ -344,6 +347,7 @@ def __init__(self, frame, buf=None, columns=None, col_space=None,
                  max_cols=None, show_dimensions=False, **kwds):
         self.frame = frame
         self.buf = buf if buf is not None else StringIO()
+        self.buf = _expand_user(self.buf)
```
Comment (jreback) [inline pandas/core/format.py]: this is I think only place you need it (covers both CSV and HTML)

File: pandas/core/format.py (hunk 2 above)
Comment (jreback) [inline pandas/core/format.py]: here ok too

**Comment by jreback** [comment]
actually on 2nd thought.

I think you these should ONLY be used in ``core/frame.py/DataFrame.to_csv and to_html``

I don't really like this here, as all of the other IO functions are localized in ``pandas/io/*`` but that's how it is for now.

**Comment by aechase** [comment]
I had the same thought, but then I assumed that would cause problems for Series.to_csv(). But, having now checked the code, I see that they are the same function, with each Series being temporarily converted to a DataFrame.

Should I put the _expand_user() calls in frame.py?

Regarding the tests: let me start out by saying that I am oh-so-close to being out of my depth here. The current suite seems to run OK (various DeprecationWarnings and RuntimeWarnings notwithstanding), but I'm a bit unclear as to how (or whether) I should create a new test function. That function would need to write and read some temporary files, but it looks like it would be difficult, if not impossible, to use testing.ensure_clean() to generate expanded and non-expanded file paths.

**Comment by jreback** [comment]
you might have a look at these: https://github.com/pydata/pandas/blob/master/pandas/io/tests/test_common.py

yeh this might be quite tricky to test

**Comment by aechase** [comment]
@jreback looks like I might have caused some trouble editing the most recent version of the release notes file...

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/ea758cd2467f3b9dbfe4aced1ce6666145c502c6

thanks!

**Pull request closed without merging**